java/code/src/com/redhat/rhn/domain/image/ImageInfoFactory.java

```diff
@@ -247,15 +272,62 @@ public static void save(ImageInfo imageInfo) {
         instance.saveObject(imageInfo);
     }
 
+    private static void removeImageFile(String path) {
+        GlobalInstanceHolder.SALT_API
+                .removeFile(Paths.get(path))
+                .orElseThrow(() -> new IllegalStateException("Can't remove image file " + path));
+    }
+
+    /**
+     * Delete a {@link DeltaImageInfo}.
+     *
+     * @param delta the delta image info to delete
+     */
+    public static void deleteDeltaImage(DeltaImageInfo delta) {
+        removeImageFile(OSImageStoreUtils.getDeltaImageFilePath(delta));
+        instance.removeObject(delta);
+    }
+
     /**
      * Delete a {@link ImageInfo}.
      *
      * @param imageInfo the image info to delete
      */
     public static void delete(ImageInfo imageInfo) {
+        imageInfo.getDeltaSourceFor().stream().forEach(ImageInfoFactory::deleteDeltaImage);
+        imageInfo.getDeltaTargetFor().stream().forEach(ImageInfoFactory::deleteDeltaImage);
+
+        imageInfo.getImageFiles().stream().forEach(f -> {
+            if (!f.isExternal()) {
+                removeImageFile(OSImageStoreUtils.getOSImageFilePath(f));
+            }
+        });
         instance.removeObject(imageInfo);
     }
 
+    /**
+     * Delete a {@link ImageInfo} and all obsolete image infos with the
+     *  same name, version and store.
+     *
+     * @param image the image info to delete
+     */
+    public static void deleteWithObsoletes(ImageInfo image) {
+        if (!image.isObsolete() && image.isBuilt()) {
```

Comment:
Why we should block this is case the image is not built? there are any side effects if we deleted it and the image is not yet built?

Reply:
Deleting current, successfully built image means that the image is no longer needed and the history can be deleted too.
Deleting failed image is normal cleanup, it should not affect the built images.